src/main/java/org/testng/internal/MethodRunner.java

```diff
@@ -0,0 +1,133 @@
+package org.testng.internal;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.testng.ITestContext;
+import org.testng.ITestResult;
+import org.testng.collections.Lists;
+import org.testng.internal.ITestInvoker.FailureContext;
+import org.testng.internal.TestMethodArguments.Builder;
+import org.testng.xml.XmlSuite;
+
+class MethodRunner implements IMethodRunner {
+
+  @Override
+  public List<ITestResult> runInSequence(TestMethodArguments arguments,
+      ITestInvoker testInvoker,
+      ITestContext context,
+      AtomicInteger invocationCount,
+      FailureContext failure,
+      Iterator<Object[]> allParameterValues,
+      boolean skipFailedInvocationCounts) {
+    List<ITestResult> result = Lists.newArrayList();
+    int parametersIndex = 0;
+    while (allParameterValues.hasNext()) {
```

Comment:
That's strange that sequence and parallel are so different. I was expecting that parallel process is just a threadpool around the sequence process.

Reply:
It doesn't seem to be the case and when I attempted something like that, its causing test failures. Now that the code beginning to get decoupled, I think we have a better chance of chasing down these sort of discrepancies.